pkg/block/s3/adapter.go

```diff
@@ -356,7 +356,22 @@ func (a *Adapter) GetPreSignedURL(ctx context.Context, obj block.ObjectPointer,
 			WithField("identifier", obj.Identifier).
 			WithError(err).Error("could not pre-sign request")
 	}
-	return preSignedURL, err
+	expiry := time.Now().Add(a.preSignedExpiry)
+	clientExpiry, clientExpiryErr := client.ExpiresAt()
+	switch {
+	case clientExpiryErr == nil:
+		if clientExpiry.Before(expiry) {
+			expiry = clientExpiry
+		}
+	case errors.Is(clientExpiryErr, ErrDoesntExpire):
+		break
+	default:
+		log.WithFields(logging.Fields{
+			"namespace":  obj.StorageNamespace,
+			"identifier": obj.Identifier,
+		}).WithError(err).Warning("Failed to get client (token) expiry: URL expiry may be too high")
+	}
+	return preSignedURL, expiry, err
```

Comment:
Just a thought: 
If the Block Adapter would have a method that returns the client expiry we could run this logic only once in the controller, I think it might make things a bit simple

Reply:
At least for S3 we support _multiple_ AWS clients, per namespace.  So this method would need to be called with a namespace parameter.  At which point it becomes simpler to push it to presigning.

Meanwhile, the controller uses this in multiple places, and we have a few more non-controller calls.  At most we'll be able to extract this logic to the adaptor level -- let's think about refactoring after adding support for Azure and GCS.